### PR TITLE
#82 - configurable scout log level

### LINKED ISSUE #82: Add Configurable LogLevel specific to Scout Agent
## The Problem

Currently, we log all scout specific messages via the $logger parameter into the agent. But that can be awkward since you can't easily set the log level distinctly between Scout and the rest of your Application.

## Proposed solution

Continue to take a $logger when constructing the agent. Output will be output there in the end.

Add a `log_level` config that takes the classic "debug", "info", "warn", "error" levels.

Add a ScoutLog class that all internal logging is routed through. It filters any messages noisier than the config that is set, and hands any left over to the provided $logger. If the config isn't set, pass through all messages. 

Change all internal logging (in this repo, and also the -laravel repo) to use the new logger.

Document the new logging configuration & how it works in https://github.com/scoutapp/slate_apm_help/blob/master/source/includes/_php.md

### PR DESCRIPTION
**Pull request by asgrim**
Fixes #82

### CODE REVIEW

File: src/CoreAgent/AutomaticDownloadAndLaunchManager.php [hunk 1]
```diff
@@ -108,7 +108,7 @@ private function run() : bool
     {
         $this->logger->debug('Core Agent Launch in Progress');
         try {
-            $logLevel   = $this->config->get('log_level');
+            $logLevel   = $this->config->get('agent_log_level');
```
Comment (asgrim) [inline src/CoreAgent/AutomaticDownloadAndLaunchManager.php]: Note: this is likely a BC break, so I'm happy to take suggestions on "naming" stuff here...

File: src/Logger/FilteredLogLevelDecorator.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    /** @var int */
+    private $minimumLogLevel;
+
+    /**
+     * @param string $minimumLogLevel e.g. `emergency`, `error`, etc. - {@see \Psr\Log\LogLevel}
+     */
+    public function __construct(LoggerInterface $realLogger, string $minimumLogLevel)
+    {
+        Assert::keyExists(self::LOG_LEVEL_ORDER, $minimumLogLevel);
```
Comment (cschneid) [inline src/Logger/FilteredLogLevelDecorator.php]: Shouldn't this get strtolower'd too?

File: src/CoreAgent/AutomaticDownloadAndLaunchManager.php (hunk 1 above)
Comment (cschneid) [reply src/CoreAgent/AutomaticDownloadAndLaunchManager.php]: The python names, and names here are mostly `core_agent_KEY` as the pattern for config. So why not `core_agent_log_level` for this.

https://docs.scoutapm.com/#core_agent_dir-1

Comment (asgrim) [reply src/CoreAgent/AutomaticDownloadAndLaunchManager.php]: OK: I've extracted all config key strings I could find and put them into a single class to keep naming consistent. I then renamed the core agent-related keys in dbbe20d95be97dc1453f50efae764eb558466504 as follows:

 - `log_level` => `core_agent_log_level`
 - `log_file` => `core_agent_log_file`
 - `config_file` => `core_agent_config_file`
 - `socket_path` => `core_agent_socket_path`
 - `download_url` => `core_agent_download_url`

These are all BC breaks of course, but if you're happy with these renames, it does add consistency. I've noted these changes in the `CHANGELOG.md`.

**Pull request merged**